### ISSUE TITLE
Security Group Audit Issues Disappear

### ISSUE THREAD

**Issue by kur3us** [Netflix/security_monkey#333]
Security Monkey sometimes indicates that a Security Group audit issue is no longer an issue.

This appears to happen when something changes in the "assigned-to" information. For example, when a Security Group is assigned to a new instance or removed from an instance, or when the tags change or are added to an instance using that Security Group.

For example, see this Security Group, which should be an audit issue:

![image](https://cloud.githubusercontent.com/assets/15004905/14973514/7d864bde-109d-11e6-9bc8-1f6dfc9cb7d1.png)

**Comment by kur3us** [comment]
I've also noticed that the issues for all accounts seem to "reset" back to being issues once every 24 hours. I've seen this consistently happen on two separate Security Monkey instances at around 10:00 AM local server time.

I've also seen the "reset" happen when an account name is changed. In this case, only the issues for that account were "reset".

**Comment by monkeysecurity** [comment]
Hey @kur3us.  An audit is run daily at 10am:

https://github.com/Netflix/security_monkey/blob/master/security_monkey/scheduler.py#L144

You wouldn't think that things would reset though.

I feel like I've seen this issue before, so if anyone has any clues or hard evidence, please add it to this ticket.  The screenshot above may simply indicate that the scheduled watcher found a change but that the auditor hadn't yet run (which should probably be changed.)  If you manually run the change watcher from the CLI, it will always audit any changed items.  The scheduled watcher should probably behave in the same way.

In any case, I'm still not sure why justifications would disappear.

**Comment by kur3us** [comment]
This issue can be easily reproduced, and I think it may be causing the problem with disappearing justifications as well (issue#339).

1. Create an EC2 instance with a Security Group containing ports open to the world (0.0.0.0/0)
2. Verify that you see it in Security Monkey with the issues noted on the open ports
3. Justify one or more of the open ports
4. Add a tag to the EC2 instance
5. Wait for Security Monkey to detect the new tag and add it to the assigned instance section of the Security Group
6. You will also see that the Security Group issues are no longer issues, and all justifications are gone as well

Screenshots below to follow the sequence of the steps above:

EC2 instance created with open ports (Time 4:55 PM):
![securitymonkeybug--1](https://cloud.githubusercontent.com/assets/15004905/15445454/bd601266-1eb3-11e6-8742-ad1914e996ae.JPG)

2 of the 4 issues justified at 5:23 PM
![securitymonkeybug--2](https://cloud.githubusercontent.com/assets/15004905/15445472/fad37188-1eb3-11e6-8daa-6b485f1181dc.JPG)

Tag added to the EC2 instance and detected by Security Monkey at 5:25 PM. The result in Security Monkey...issues disappeared,..justifications disappeared:
![securitymonkeybug--3](https://cloud.githubusercontent.com/assets/15004905/15445487/3efb6f0a-1eb4-11e6-9ea7-1e746dfad9b4.JPG)

(New tag can be seen in green below: "SecurityMonkeyTest".)

![securitymonkeybug--4](https://cloud.githubusercontent.com/assets/15004905/15445488/476827be-1eb4-11e6-922b-d6a69399fcd5.JPG)

And if we were to wait until 10:00 AM local on the server, we will likely see this turn into an issue again, but the justifications will still be gone.

**Comment by monkeysecurity** [comment]
Thanks for finding a way to reproduce.  That will help in tracking this down.

**Comment by monkeysecurity** [comment]
When an issue is "fixed", security monkey currently deletes the item from the issue table. Instead, it should probably just mark the issue as fixed by adding a boolean column.  Then, if the issue later reappears, the justification would still exist.

I'm seeing this bug a lot lately as some of the AWS APIs have been incorrectly telling me that an item (s3 bucket or iam user login profile or something like that) has been deleted, then it reappears a few minutes later.  The issue/justification are then deleted.

This issue needs to be fixed before we have our next release.

**Comment by monkeysecurity** [comment]
I accidentally closed this issue just a moment ago.  Reopening.

**Comment by monkeysecurity** [comment]
@kur3us Does this still occur if you set `SECURITYGROUP_INSTANCE_DETAIL = "NONE"` ?

**Comment by monkeysecurity** [comment]
Are there any exceptions or stack traces in the logs during this time frame?

**Comment by monkeysecurity** [comment]
When you renamed an account, did you also restart the scheduler?

**Comment by kur3us** [comment]
Sorry for the delay in responding. I just switched SECURITYGROUP_INSTANCE_DETAIL to NONE. I'll check and let you know if this changes anything.

I'll check the logs for stack traces as well. 

I didn't rename an account. This issue happens when instance information changes with an instance associated with the Security Group in question. See the steps to reproduce in my comment on May 20th.

**Comment by cstewart87** [comment]
@monkeysecurity - I believe we recently came across this issue (or similar) as well.  We noticed it when some of our justifications were being blown away.  Upon further investigation, we realized that any item change (ephemeral and non-ephemeral), would cause all audit issues to be removed, then re-created during the next audit.

From what I can tell, this is due to the code at https://github.com/Netflix/security_monkey/blob/develop/security_monkey/watcher.py#L446 .  Based on testing, when an item changes, the `valid_item.audit_issues` is always an empty array.

IMO, in order to fix this you would want to set audit_issues to `old_item.audit_issues`, if any exist.

Thoughts?
